File: 678.go
func checkValidString(s string) bool {
    cmin, cmax := 0, 0

    for _, c := range s {
        if c == '(' {
            cmin++
            cmax++
        } else if c == ')' {
            cmin = max(0, cmin - 1)
            cmax--
        } else {
            cmin = max(0, cmin - 1)
            cmax++
        }

        if cmax < 0 {
            return false
        }
    }

    return cmin == 0
}

File: 678.cpp
class Solution {
public:
    bool checkValidString(string s) {
        int cmin = 0;
        int cmax = 0;

        for (const auto& c : s) {
            if (c == '(') {
                cmin++;
                cmax++;
            } else if (c == ')') {
                cmin = max(0, cmin - 1);
                cmax--;
            } else {
                cmin = max(0, cmin - 1);
                cmax++;
            }

            if (cmax < 0) {
                return false;
            }
        }

        return cmin == 0;
    }
};

